File: Arrays/Easy/Reverse Array/ReverseArray.java

public class ReverseArray {
    // Approach 1 :- Using another array..
    // Time Complexity:- O(n), n is the size of vector.
    // Space Complexity:- O(n)
    static void reverse_Approach_1(int arr[]) {

        // size of array.
        int n = arr.length;

        // copy all the elements of arr into aux array in reverse order. 
        int aux[] = new int[n];
        int k = 0;
        for (int i = n-1; i >= 0 ; i--) {
            aux[k++] = arr[i];
        }

        // copy back all element of aux array into original array.
        for (int i = 0; i < n ; i++) {
            arr[i] = aux[i];
        }
    }

    // Approach 2 :- Using 2 pointers.
    // Time Complexity:- O(n), n is the size of vector.
    // Space Complexity:- O(1)
    static void reverse_Approach_2(int arr[]) {
        int start = 0;
        int end = arr.length - 1;

        while (start < end)
        {
            // swap arr[start], arr[end]
            arr[start] = arr[start] ^ arr[end];
            arr[end] = arr[start] ^ arr[end];
            arr[start] = arr[start] ^ arr[end];

            // update pointers
            start++;
            end--;
        }
    }

    public static void main(String[] args) {
        int arr[] = { 1, 2, 3, 4, 5};

        // reverse_Approach_1(arr);
        reverse_Approach_2(arr);
        for(int x : arr)
        {
            System.out.print(x + " ");
        }
    }
}

File: 1. Arrays/Easy/Reverse Array/ReverseArray.cpp
#include <bits/stdc++.h>
using namespace std;
#include <bits/stdc++.h>
using namespace std;

/*------------------------------ Solution Class -----------------------------*/
class Solution
{
public:
    // Approach 1 :- Using another array..
    // Time Complexity:- O(n), n is the size of vector.
    // Space Complexity:- O(n)
    void reverse_Approach_1(vector<int> &nums)
    {
        // copy all the elements of "nums" into "arr" array in reverse order. 
        vector<int> arr;
        for (int i = nums.size() - 1; i >= 0; i--)
        {
            arr.push_back(nums[i]);
        }

        // copy back all element of arr into original array.
        for (int i = 0; i < arr.size(); i++)
        {
            nums[i] = arr[i];
        }

        
    }
    // Approach 2 :- Using 2 pointers.
    // Time Complexity:- O(n), n is the size of vector.
    // Space Complexity:- O(1)
    void reverse_Approach_2(vector<int> &nums)
    {
        int start = 0;
        int end = nums.size() - 1;

        while (start <= end)
        {
            swap(nums[start++], nums[end--]);
        }
    }

    /* Method to print vector. */
    void print(vector<int> arr)
    {
        for (int i = 0; i < arr.size(); i++)
        {
            cout << arr[i] << " ";
        }
        cout << endl;
    }
};
/*-------------------------- Solution Class Ends Here. ------------------------- */
int main()
{
    vector<int> arr = {1, 2, 3, 4, 5};

    Solution s;
    s.reverse_Approach_1(arr);
    s.print(arr);
}
